src/test/spec/unified_runner/mod.rs

```diff
@@ -0,0 +1,370 @@
+            }
+        }
+
+        if let Some(ref create_entities) = test_file.create_entities {
+            test_runner.populate_entity_map(create_entities).await;
+        }
+
+        for operation in test_case.operations {
+            match operation.object {
+                OperationObject::TestRunner => {
+                    operation
+                        .execute_test_runner_operation(&mut test_runner)
+                        .await;
+                }
+                OperationObject::Entity(ref id) => {
+                    let result = operation
+                        .execute_entity_operation(id, &mut test_runner)
+                        .await;
+
+                    if let Some(ref id) = operation.save_result_as_entity {
+                        match &result {
+                            Ok(Some(entity)) => {
+                                if test_runner
+                                    .entities
+                                    .insert(id.clone(), entity.clone())
+                                    .is_some()
+                                {
+                                    panic!("Entity with id {} already present in entity map", id);
+                                }
+                            }
+                            Ok(None) => panic!("{} did not return an entity", operation.name),
+                            Err(_) => panic!("{} should succeed", operation.name),
+                        }
+                    }
+
+                    if let Some(ref expect_result) = operation.expect_result {
+                        let result = result
+                            .unwrap_or_else(|_| panic!("{} should succeed", operation.name))
+                            .unwrap_or_else(|| {
+                                panic!("{} should return an entity", operation.name)
+                            });
+                        match result {
+                            Entity::Bson(ref result) => {
+                                assert!(results_match(
+                                    Some(result),
+                                    &expect_result,
+                                    operation.returns_root_documents(),
+                                    Some(&test_runner.entities),
+                                ));
+                            }
+                            _ => panic!(
+                                "Incorrect entity type returned from {}, expected BSON",
+                                operation.name
+                            ),
+                        }
+                    } else if let Some(expect_error) = operation.expect_error {
+                        let error = result
+                            .expect_err(&format!("{} should return an error", operation.name));
+                        expect_error.verify_result(error);
+                    }
+                }
+            }
+        }
+
+        test_runner.disable_fail_points().await;
```

Comment:
If a test panics before this is run, the failpoints won't be disabled. To deal with that, we can use the (newly introduced since this PR was opened) `FailPoint` and `FailPointGuard` types by storing the `FailPointGuard`s instead of just the disable-failpoint-command documents. Their `Drop` implementations (which disable the failpoints) should get invoked even in the event of a panic. In the normal case, we could just drop them all manually via `disable_fail_points`.

As part of that, we'd also need a `Deserialize` implementation for `FailPoint`, but since that type is just a wrapper around a document that should be pretty easy. Also, when using the `FailPointGuard` type, the tokio tests need to use the threaded scheduler since the `Drop` implementations use `RUNTIME.block_on` ([example](https://github.com/mongodb/mongo-rust-driver/blob/master/src/test/spec/retryable_reads.rs#L5)).